aiida/restapi/resources.py

```diff
@@ -207,6 +208,130 @@ def get(self, id=None, page=None):  # pylint: disable=redefined-builtin,invalid-
         return self.utils.build_response(status=200, headers=headers, data=data)
 
 
+class QueryBuilder(BaseResource):
+    """
+    Representation of a QueryBuilder REST API resource (instantiated with a queryhelp JSON).
+
+    It supports POST requests taking in JSON :py:func:`~aiida.orm.querybuilder.QueryBuilder.queryhelp`
+    objects and returning the :py:class:`~aiida.orm.querybuilder.QueryBuilder` result accordingly.
+    """
+    from aiida.restapi.translator.nodes.node import NodeTranslator
+
+    _translator_class = NodeTranslator
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        # HTTP Request method decorators
+        if 'get_decorators' in kwargs and isinstance(kwargs['get_decorators'], (tuple, list, set)):
+            self.method_decorators.update({'post': list(kwargs['get_decorators'])})
+
+    def get(self):  # pylint: disable=arguments-differ
+        """Static return to state information about this endpoint."""
+        data = {
+            'message': (
+                'Method Not Allowed. Use HTTP POST requests to use the AiiDA QueryBuilder. '
+                'POST JSON data, which MUST be a valid QueryBuilder.queryhelp dictionary as a JSON object. '
+                'See the documentation at https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/'
+                'database.html?highlight=QueryBuilder#the-queryhelp for more information.'
+            ),
+        }
+
+        headers = self.utils.build_headers(url=request.url, total_count=1)
+        return self.utils.build_response(
+            status=405,  # Method Not Allowed
+            headers=headers,
+            data={
+                'method': request.method,
+                'url': unquote(request.url),
+                'url_root': unquote(request.url_root),
+                'path': unquote(request.path),
+                'query_string': request.query_string.decode('utf-8'),
+                'resource_type': self.__class__.__name__,
+                'data': data,
+            },
+        )
+
+    def post(self):
+        """
+        POST method to pass query help JSON.
+
+        This assumes the posted data is a complete and valid QueryBuilder queryhelp JSON object.
+        We could create a RequestParser that completely redefines all valid keys in the queryhelp,
+        but for now, it's much easier to simply pass-through the posted queryhelp and make sure
+        it can be handled by the built-in NodeTranslator.
+
+        This uses the NodeTranslator in order to best return Nodes according to the general AiiDA
+        REST API data format, while still allowing the return of other AiiDA entities.
+
+        :return: QueryBuilder result of AiiDA entities in "standard" REST API format.
+        """
+        # pylint: disable=protected-access
+        self.trans._query_help = json.loads(request.json) if isinstance(request.json, str) else request.json
```

Comment:
This is not the `requests` library, but rather a flask `Request` object.

Reply:
does the format of `.json` depend on the version of flask?

Reply:
Probably not. I'm just trying to avoid `bytes` and `None` here, since `json.loads()` only takes `str` as far as I know?

Reply:
Oh, ok. To me it looked like you're expecting `request.json` to either be a string or already a dictionary.

If it is `bytes`, I guess you will either need a different way to load the dictionary from it or you could raise an error.

The check for `None` would be great to make explicit.

Reply:
> Oh, ok. To me it looked like you're expecting `request.json` to either be a string or already a dictionary.

Ah, you're right. I think the `json` attribute for a flask `Request` can be either a `str` or the parsed JSON to Python?

> The check for `None` would be great to make explicit.

Right.

Reply:
Here's the documentation of the relevant function
https://flask.palletsprojects.com/en/1.1.x/api/#flask.Request.get_json

It seems to me that you can get the behaviour intended (try parsing even if content type is not `json`) by using `get_json(force=True)`.
If parsing fails, it will raise a `BadRequest` error automatically; you'll still need to check for `None`.